training_scripts/plot_nnunet_training_log.py

```diff
@@ -0,0 +1,157 @@
+    """
+    fold_pattern = re.compile(r'Desired fold for training: (\d+)')
+    epoch_pattern = re.compile(r'Epoch (\d+)')
+    dice_pattern = re.compile(r'Pseudo dice \[([^,\]]+(?:, [^,\]]+)*)\]')
+
+    with open(log_file_path, 'r') as file:
+        lines = file.readlines()
+
+    epoch_and_dice_data = []
+    for line in lines:
+        fold_match = fold_pattern.search(line)
+        epoch_match = epoch_pattern.search(line)
+        dice_match = dice_pattern.search(line)
+
+        if fold_match:
+            fold_number = int(fold_match.group(1))
+
+        if epoch_match:
+            epoch = int(epoch_match.group(1))
+            epoch_and_dice_data.append({'epoch': epoch, 'pseudo_dice': None})
+
+        elif dice_match:
+            # Extracting the list using regular expression
+            list_match = re.search(r'\[.*\]', dice_match.group())
+
+            if list_match:
+                extracted_list_str = list_match.group(0)
+                # Replace 'nan' with the actual nan value
+                extracted_list_str = extracted_list_str.replace('nan', 'float("nan")')
+                # Evaluating the string representation of the list
+                extracted_list = eval(extracted_list_str)
+                epoch_and_dice_data[-1]['pseudo_dice'] = extracted_list
+
+    # Check if fold_number is instance of int ("if not fold_number:" does not work for '0')
+    if not isinstance(fold_number, int):
+        fold_number = 'all'
+
+    return epoch_and_dice_data, fold_number
+
+
+def main():
+
+    # Parse the command line arguments
+    parser = get_parser()
+    args = parser.parse_args()
+
+    # Get absolute path to the log file
+    log_file_path = os.path.abspath(os.path.expanduser(args.i))
+
+    data, fold_number = extract_epoch_and_dice(log_file_path)
+
+    # Convert data to a Pandas DataFrame
+    df = pd.DataFrame(data)
+
+    # Create columns for each element in pseudo_dice
+    # [:-1] is used to remove the last row, which is empty
+    df_pseudo_dice = pd.DataFrame(df['pseudo_dice'].to_list()[:-1],
+                                  columns=[f'pseudo_dice_{i+1}' for i in range(len(df['pseudo_dice'].iloc[0]))])
+
+    # Concatenate the new DataFrame with the original DataFrame
+    df = pd.concat([df, df_pseudo_dice], axis=1).drop('pseudo_dice', axis=1)
+
+    # Compute mean of pseudo dice across all classes
+    df['pseudo_dice_mean'] = df.iloc[:, 1:].mean(axis=1)
+
+    # Plotting using Plotly Express
+    fig = px.line(df, x='epoch', y=df.columns[1:], title='Pseudo Dice vs. Epoch')
+    # Update the line color for 'pseudo_dice_mean' to black
+    fig.update_traces(line=dict(color='black', width=3), selector=dict(name='pseudo_dice_mean'))
+    # Fix the y-axis range to be between 0 and 1
+    fig.update_yaxes(range=[-0.1, 1.1])
+    # Add y-axis title
+    fig.update_yaxes(title_text='Dice')
+    # Add title with fold number
+    fig.update_layout(title=f'Fold {fold_number} -- Pseudo Dice vs. Epoch (Training)')
```

Comment:
Good catch! You are right! 

Indeed, based on the nnUNet [documentation](https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/how_to_use_nnunet.md), it should be `(Validation)`.

> This approximation is the average Dice score of the foreground classes. It needs to be taken with a big (!) grain of salt because it is computed on randomly drawn patches from the validation data at the end of each epoch

So this just confirms that for `fold_all`, the pseudo dice is completely irrelevant because it is computed on training samples (because there are no validation samples), right?

Reply:
yes!

Reply:
Fixed to `Validation` in https://github.com/ivadomed/utilities/pull/41/commits/d0a651e6451c44efe17747b282fd061b266bc0f0